### PR TITLE
Documentation

### LINKED ISSUE #100: Document code
Add some more documentation, This should make it more easy for new people to get into the code.

### PR DESCRIPTION
**Pull request by scruplelesswizard**
Fixes #100

### CODE REVIEW

**Comment by lbb** [comment]
This resolves #100

File: handler/routing.go [hunk 1]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
 		Handler(b.authenticated(b.deleteNodeHandler()))
 }
 
+// datacentersHandler serves a list of datacenters.
+// Admin only!
 func (b Routing) datacentersHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 2]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
 		decodeDatacentersReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// datacenterHandler server information for a datacenter.
+// Admin only!
 func (b Routing) datacenterHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from b to r or to a full word

File: handler/routing.go [hunk 3]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		datacenterEndpoint(b.datacenters, b.kubernetesProviders, b.cloudProviders),
 		decodeDcReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// newClusterHandler creates a new cluster.
 func (b Routing) newClusterHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 4]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		newClusterEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeNewClusterReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// clusterHandler returns a cluster object.
 func (b Routing) clusterHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 5]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		clusterEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeClusterReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// setCloudHandler updates a cluster.
 func (b Routing) setCloudHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 6]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		setCloudEndpoint(b.datacenters, b.kubernetesProviders, b.cloudProviders),
 		decodeSetCloudReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// kubeconfigHandler returns the cubeconfig for the cluster.
 func (b Routing) kubeconfigHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 7]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		kubeconfigEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeKubeconfigReq,
 		encodeKubeconfig,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// clustersHandler lists all clusters from a user.
 func (b Routing) clustersHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 8]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		clustersEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeClustersReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// deleteClusterHandler deletes a cluster.
 func (b Routing) deleteClusterHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 9]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		deleteClusterEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeDeleteClusterReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// nodesHandler returns all nodes from a user.
 func (b Routing) nodesHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 10]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		nodesEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeNodesReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// createNodesHandler let's you create nodes.
 func (b Routing) createNodesHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: handler/routing.go [hunk 11]
```diff
@@ -112,120 +112,133 @@ func (b Routing) Register(mux *mux.Router) {
...
+		createNodesEndpoint(b.kubernetesProviders, b.cloudProviders),
 		decodeCreateNodesReq,
 		encodeJSON,
 		httptransport.ServerErrorLogger(b.logger),
 		defaultHTTPErrorEncoder(),
 	)
 }
 
+// deleteNodeHandler let's you delete nodes.
 func (b Routing) deleteNodeHandler() http.Handler {
```
Comment (scruplelesswizard) [inline handler/routing.go]: The routing should be renamed, either from `b` to `r` or to a full word

File: provider/cloud/bringyourown/provider.go [hunk 12]
```diff
@@ -25,7 +25,7 @@ func NewCloudProvider() provider.CloudProvider {
 	return &bringyourown{}
 }
 
-func (b *bringyourown) CreateAnnotations(cloud *api.CloudSpec) (map[string]string, error) {
+func (b *bringyourown) Marshal(cloud *api.CloudSpec) (map[string]string, error) {
```
Comment (scruplelesswizard) [inline provider/cloud/bringyourown/provider.go]: Marshal what?

File: provider/cloud/bringyourown/provider.go [hunk 13]
```diff
@@ -39,7 +39,7 @@ func (b *bringyourown) CreateAnnotations(cloud *api.CloudSpec) (map[string]strin
 	return as, nil
 }
 
-func (b *bringyourown) Cloud(as map[string]string) (*api.CloudSpec, error) {
+func (b *bringyourown) Unmarshal(as map[string]string) (*api.CloudSpec, error) {
```
Comment (scruplelesswizard) [inline provider/cloud/bringyourown/provider.go]: Unmarshal what?

File: provider/cloud/digitalocean/provider.go [hunk 14]
```diff
@@ -35,15 +35,15 @@ func NewCloudProvider(dcs map[string]provider.DatacenterMeta) provider.CloudProv
 	}
 }
 
-func (do *digitalocean) CreateAnnotations(cloud *api.CloudSpec) (map[string]string, error) {
+func (do *digitalocean) Marshal(cloud *api.CloudSpec) (map[string]string, error) {
```
Comment (scruplelesswizard) [inline provider/cloud/digitalocean/provider.go]: Marshal what?

File: provider/cloud/digitalocean/provider.go [hunk 15]
```diff
@@ -35,15 +35,15 @@ func NewCloudProvider(dcs map[string]provider.DatacenterMeta) provider.CloudProv
...
 	as := map[string]string{
 		tokenAnnotationKey:  cloud.Digitalocean.Token,
 		sshKeysAnnotionsKey: strings.Join(cloud.Digitalocean.SSHKeys, ","),
 	}
 	return as, nil
 }
 
-func (do *digitalocean) Cloud(annotations map[string]string) (*api.CloudSpec, error) {
+func (do *digitalocean) Unmarshal(annotations map[string]string) (*api.CloudSpec, error) {
```
Comment (scruplelesswizard) [inline provider/cloud/digitalocean/provider.go]: Unmarshal what?

File: provider/cloud/fake/provider.go [hunk 16]
```diff
@@ -21,14 +21,14 @@ func NewCloudProvider() provider.CloudProvider {
 	return &fakeCloudProvider{}
 }
 
-func (p *fakeCloudProvider) CreateAnnotations(cloud *api.CloudSpec) (map[string]string, error) {
+func (p *fakeCloudProvider) Marshal(cloud *api.CloudSpec) (map[string]string, error) {
```
Comment (scruplelesswizard) [inline provider/cloud/fake/provider.go]: Marshal what?

File: provider/cloud/fake/provider.go [hunk 17]
```diff
@@ -21,14 +21,14 @@ func NewCloudProvider() provider.CloudProvider {
...
 	as := map[string]string{}
 	as[tokenAnnotationKey] = cloud.Fake.Token
 
 	return as, nil
 }
 
-func (p *fakeCloudProvider) Cloud(as map[string]string) (*api.CloudSpec, error) {
+func (p *fakeCloudProvider) Unmarshal(as map[string]string) (*api.CloudSpec, error) {
```
Comment (scruplelesswizard) [inline provider/cloud/fake/provider.go]: Unmarshal what?

File: provider/kubernetes/marshal.go [hunk 18]
```diff
@@ -245,40 +245,40 @@ func MarshalCluster(cps map[string]provider.CloudProvider, c *api.Cluster, ns *k
...
 
-	return as, nil
+	return annotations, nil
 }
 
 // unmarshalClusterCloud sets the Spec.Cloud field according to the annotations.
-func unmarshalClusterCloud(cpName string, cp provider.CloudProvider, as map[string]string) (*api.CloudSpec, error) {
+func unmarshalClusterCloud(cpName string, cp provider.CloudProvider, annotations map[string]string) (*api.CloudSpec, error) {
 	prefix := cloudProviderAnnotationPrefix(cpName)
 	cloudAs := map[string]string{}
```
Comment (scruplelesswizard) [inline provider/kubernetes/marshal.go:269]: Might as well make this `cloudAnnotations`

File: provider/types.go [hunk 19]
```diff
@@ -23,9 +23,10 @@ type User struct {
 
 // CloudSpecProvider declares methods for converting a cloud spec to/from annotations.
 type CloudSpecProvider interface {
-	PrepareCloudSpec(*api.Cluster) error
-	CreateAnnotations(*api.CloudSpec) (map[string]string, error)
-	Cloud(annotations map[string]string) (*api.CloudSpec, error)
+	// Initialises fields needed for cloudSpec
```
Comment (scruplelesswizard) [inline provider/types.go]: Noisy. Please remove

File: provider/types.go [hunk 20]
```diff
@@ -44,10 +45,19 @@ type CloudProvider interface {
 
 // KubernetesProvider declares the set of methods for interacting with a Kubernetes cluster.
 type KubernetesProvider interface {
+	// NewCluster creates a cluster under the provided user with the given ClusterSpec prerequisites.
 	NewCluster(user User, cluster string, spec *api.ClusterSpec) (*api.Cluster, error)
+
+	// Cluster return a Cluster struct from the namespace, given the user and cluster.
 	Cluster(user User, cluster string) (*api.Cluster, error)
+
+	// SetCloud updates a given cluster from a user with the CloudSpec settings.
```
Comment (scruplelesswizard) [inline provider/types.go]: // SetCloud updates _CloudSpec settings on the_ given cluster _for_ the given user

File: provider/types.go [hunk 21]
```diff
@@ -44,10 +45,19 @@ type CloudProvider interface {
 
 // KubernetesProvider declares the set of methods for interacting with a Kubernetes cluster.
 type KubernetesProvider interface {
+	// NewCluster creates a cluster under the provided user with the given ClusterSpec prerequisites.
 	NewCluster(user User, cluster string, spec *api.ClusterSpec) (*api.Cluster, error)
+
+	// Cluster return a Cluster struct from the namespace, given the user and cluster.
```
Comment (scruplelesswizard) [inline provider/types.go]: // Cluster return a Cluster struct  ~~from the namespace~~, given the user and cluster.

File: provider/types.go [hunk 22]
```diff
@@ -44,10 +45,19 @@ type CloudProvider interface {
 
 // KubernetesProvider declares the set of methods for interacting with a Kubernetes cluster.
 type KubernetesProvider interface {
+	// NewCluster creates a cluster under the provided user with the given ClusterSpec prerequisites.
```
Comment (scruplelesswizard) [inline provider/types.go]: // NewCluster creates a cluster _for_ the provided user _using_ the given ClusterSpec

File: provider/types.go [hunk 23]
```diff
@@ -44,10 +45,19 @@ type CloudProvider interface {
...
 	NewCluster(user User, cluster string, spec *api.ClusterSpec) (*api.Cluster, error)
+
+	// Cluster return a Cluster struct from the namespace, given the user and cluster.
 	Cluster(user User, cluster string) (*api.Cluster, error)
+
+	// SetCloud updates a given cluster from a user with the CloudSpec settings.
 	SetCloud(user User, cluster string, cloud *api.CloudSpec) (*api.Cluster, error)
+
+	// Cluster returns all cluster for a given user.
```
Comment (scruplelesswizard) [inline provider/types.go]: // Cluster returns all cluster_s_ for a given user.

File: types.go [hunk 24]
```diff
@@ -11,15 +11,21 @@ type Metadata struct {
 	UID      string `json:"uid,omitempty"`
 
 	// private fields
+	// Annotations are kubernetes Annotations from the namespace.
+	// Used to store data persistend.
```
Comment (scruplelesswizard) [inline types.go]: _// Annotations represent Annotations on Kubernetes Namespace for the respective cluster, which are used to store persistent data for the cluster._

File: types.go [hunk 25]
```diff
@@ -11,15 +11,21 @@ type Metadata struct {
...
 
 // DigitaloceanNodeSpec specifies a digital ocean node.
 type DigitaloceanNodeSpec struct {
-	Type    string   `json:"type"`
-	Size    string   `json:"size"`
-	SSHKeys []string `json:"sshKeys,omitempty"`
+	// Type specifies the type of the image used to create the node.
+	Type string `json:"type"`
+	// Size is the size of the node.
+	Size string `json:"size"`
```
Comment (scruplelesswizard) [inline types.go:25]: What does `Size` represent? Disk size? Procs? Mem? The instance type? 
`Size` here is very ambiguous, and not clear across providers

File: types.go [hunk 26]
```diff
@@ -11,15 +11,21 @@ type Metadata struct {
...
 	Annotations map[string]string `json:"-"`
 	User        string            `json:"-"`
 }
 
 // DigitaloceanNodeSpec specifies a digital ocean node.
 type DigitaloceanNodeSpec struct {
-	Type    string   `json:"type"`
-	Size    string   `json:"size"`
-	SSHKeys []string `json:"sshKeys,omitempty"`
+	// Type specifies the type of the image used to create the node.
```
Comment (scruplelesswizard) [inline types.go]: What does `Type` represent? Proc architecture? OS? `Type`here is very ambiguous, and not clear across providers

File: types.go [hunk 27]
```diff
@@ -11,15 +11,21 @@ type Metadata struct {
...
+	Type string `json:"type"`
+	// Size is the size of the node.
+	Size string `json:"size"`
+	// SSHKeyNames represent the fingerprints of the keys.
+	// Digitalocean utilizes the fingerprints to identify SSHKeys stored at digitalocean.
```
Comment (scruplelesswizard) [inline types.go]: // Digital_O_cean utilizes fingerprints to identify _public_ SSH Keys within the DigitalOcean platform.

File: types.go [hunk 28]
```diff
@@ -34,7 +40,9 @@ type FakeNodeSpec struct {
 
 // NodeSpec mutually stores data of a cloud specific node.
 type NodeSpec struct {
-	DC           string                `json:"dc"`
+	// DC contains the name of the datacenter the node is located in.
+	DC string `json:"dc"`
```
Comment (scruplelesswizard) [inline types.go]: s/DC/DataCenterName/

File: types.go [hunk 29]
```diff
@@ -55,7 +63,9 @@ type Node struct {
 
 // DigitaloceanCloudSpec specifies access data to digital ocean.
 type DigitaloceanCloudSpec struct {
-	Token   string   `json:"token"`
+	// Token grants access to the digitalocean api.
+	Token string `json:"token"`
```
Comment (scruplelesswizard) [inline types.go:68]: `APIToken`

File: types.go [hunk 30]
```diff
@@ -55,7 +63,9 @@ type Node struct {
 
 // DigitaloceanCloudSpec specifies access data to digital ocean.
 type DigitaloceanCloudSpec struct {
-	Token   string   `json:"token"`
+	// Token grants access to the digitalocean api.
```
Comment (scruplelesswizard) [inline types.go]: _// APIToken is used to authenticate with the DigitalOcean API._

File: types.go [hunk 31]
```diff
@@ -11,15 +11,21 @@ type Metadata struct {
...
 // DigitaloceanNodeSpec specifies a digital ocean node.
 type DigitaloceanNodeSpec struct {
-	Type    string   `json:"type"`
-	Size    string   `json:"size"`
-	SSHKeys []string `json:"sshKeys,omitempty"`
+	// Type specifies the type of the image used to create the node.
+	Type string `json:"type"`
+	// Size is the size of the node.
+	Size string `json:"size"`
+	// SSHKeyNames represent the fingerprints of the keys.
```
Comment (scruplelesswizard) [inline types.go]: // SSHKeyFingerprints represent the fingerprints of the keys.

File: types.go [hunk 32]
```diff
@@ -11,15 +11,21 @@ type Metadata struct {
...
+	// Digitalocean utilizes the fingerprints to identify SSHKeys stored at digitalocean.
+	SSHKeyNames []string `json:"sshKeys,omitempty"`
```
Comment (scruplelesswizard) [inline types.go]: `SSHKeyNames` is not an accurate descrition, please use `SSHKeyFingerprints` instead

File: types.go [hunk 33]
```diff
@@ -55,7 +63,9 @@ type Node struct {
 
 // DigitaloceanCloudSpec specifies access data to digital ocean.
 type DigitaloceanCloudSpec struct {
-	Token   string   `json:"token"`
+	// Token grants access to the digitalocean api.
+	Token string `json:"token"`
+	// SSHKeys are real SSH keys ("id-alg .....") used by the cloud-init.
```
Comment (scruplelesswizard) [inline types.go]: _// SSHKeys are SSH keys used in the cloud-init generation to deploy to nodes._

File: types.go [hunk 34]
```diff
@@ -72,21 +82,28 @@ type FakeCloudSpec struct {
 
 // FlannelNetworkSpec specifies a deployed flannel network.
 type FlannelNetworkSpec struct {
+	// CIDR is the subnet used by flannel.
+	// It need's to be provided in CIDR notation.
```
Comment (scruplelesswizard) [inline types.go]: ~~// It need's to be provided in CIDR notation.~~

File: types.go [hunk 35]
```diff
@@ -72,21 +82,28 @@ type FakeCloudSpec struct {
 
 // FlannelNetworkSpec specifies a deployed flannel network.
 type FlannelNetworkSpec struct {
+	// CIDR is the subnet used by flannel.
```
Comment (scruplelesswizard) [inline types.go]: // CIDR is the subnet used by Flannel in CIDR notation.

File: types.go [hunk 36]
```diff
@@ -72,21 +82,28 @@ type FakeCloudSpec struct {
...
+	// It need's to be provided in CIDR notation.
+	// See RFC: 4632, eg. "127.1.0.0/16"
```
Comment (scruplelesswizard) [inline types.go]: _// e.g. "127.1.0.0/16"_

File: types.go [hunk 37]
```diff
@@ -72,21 +82,28 @@ type FakeCloudSpec struct {
...
 	CIDR string
 }
 
 // NetworkSpec specifies the deployed network.
 type NetworkSpec struct {
+	// FlannelNetworkSpec holds the informations needed for running a flannel network.
```
Comment (scruplelesswizard) [inline types.go]: Information is never plural. 
Flannel is a proper noun (name) and should be capitalized 
_// FlannelNetworkSpec holds the required information to configure Flannel_

File: types.go [hunk 38]
```diff
@@ -72,21 +82,28 @@ type FakeCloudSpec struct {
...
 	Flannel FlannelNetworkSpec
 }
 
 // CloudSpec mutually stores access data to a cloud provider.
 type CloudSpec struct {
-	DC           string                 `json:"dc"`
+	// The datacenter the users 'cloud' lives in.
```
Comment (scruplelesswizard) [inline types.go]: // _DC represents_ the datacenter the user_'_s 'cloud' lives in.

File: types.go [hunk 39]
```diff
@@ -72,21 +82,28 @@ type FakeCloudSpec struct {
...
+	DC string `json:"dc"`
+	// Network holds the network specifications.
```
Comment (scruplelesswizard) [inline types.go]: Network spec for what?

File: types.go [hunk 40]
```diff
@@ -158,8 +175,9 @@ type ClusterStatus struct {
 
 // ClusterSpec specifies the data for a new cluster.
 type ClusterSpec struct {
-	Cloud             *CloudSpec `json:"cloud,omitempty"`
-	HumanReadableName string     `json:"humanReadableName"`
+	Cloud *CloudSpec `json:"cloud,omitempty"`
+	// The name choosen by the user.
```
Comment (scruplelesswizard) [inline types.go]: _// HumanReadableName is the cluster name provided by the user_

**Review by scruplelesswizard — COMMENTED** [review]
@realfake Please resolve the listed issues. I cannot request changes as this is set as 'my' PR

File: provider/cloud/bringyourown/provider.go (hunk 12 above)
Comment (lbb) [reply provider/cloud/bringyourown/provider.go]: Marshal fits part of an interface. CloudSpecProvider, it's not the go way to be repetitive. `CloudSpecProvider.MarshalCloud` wouldn't make sense. The definition is part of the function prototype.

Comment (scruplelesswizard) [reply provider/cloud/bringyourown/provider.go]: In this case you are not marshaling the struct this is being called on, rather you are marshalling the struct _and_ the `CloudSpec`, so this should be called `MarshalCloudSpec`. The current definition of Marshal implies that you are only operating on the struct

**Comment by scruplelesswizard** [comment]
@realfake What is the status on this? Can we fix the conflicts, clean up the nits and get it merged in?

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**